### ISSUE TITLE
elf: An issue with extracting kernel version on Debian distributions

### ISSUE THREAD

**Issue by denisgersh** [iovisor/gobpf#158]
Debian's kernel version extraction doesn't reach on Debian distributions. There is an issue with extraction logic in [CurrentKernelVersion](https://github.com/iovisor/gobpf/blob/8f8e0b6e4fcbbda9274a9fbfa8751e1eb88cd5d3/elf/kernel_version.go#L94) function. If we can extract the kernel version with Ubuntu's way, then we skip the Debian's one. It's not working on Debian distribution.

As an example, I tried to load a BPF compiled object with a simple kprobe event and undefined kernel version inside it on Debian 9 - I got an **EINVAL** error.

_(Linux debian-gnu-linux-vm 4.9.0-8-amd64 #1 SMP Debian 4.9.110-3+deb9u6 (2018-10-08) x86_64 GNU/Linux)_

As a solution, just need to change the order of calling functions in [CurrentKernelVersion](https://github.com/iovisor/gobpf/blob/8f8e0b6e4fcbbda9274a9fbfa8751e1eb88cd5d3/elf/kernel_version.go#L94) function to:

```
func CurrentKernelVersion() (uint32, error) {
	version, err = currentVersionDebian() // It was after Ubuntu's function
	if err == nil {
		return version, nil
	}

        version, err := currentVersionUbuntu()
	if err == nil {
		return version, nil
	}

	return currentVersionUname()
}
```

Also, the regex of extracting kernel version in Debian has an issue.

After changing the regex in [kernel_version.go](https://github.com/iovisor/gobpf/blob/8f8e0b6e4fcbbda9274a9fbfa8751e1eb88cd5d3/elf/kernel_version.go#L78) function to:
```regexp.MustCompile(`.* SMP Debian (\d+\.\d+.\d+-\d+) *.`)```

Then the regex passed correctly and a correct kernel version was extracted.

**Comment by andreynkravtsov** [comment]
Just ran into this exact issue, thanks for the workaround. It's unfortunate nobody patched the project yet for same

**Comment by andreynkravtsov** [comment]
Actually, in my case the regex would need to change a bit

My /proc/version is
Linux version 4.9.0-11-amd64 (debian-kernel@lists.debian.org) (gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1) ) #1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11)

Given that we only care about the 4.9.189 in this sequence, I think the more appropriate regex is var debianVersionRegex = regexp.MustCompile(`.* SMP Debian (\d+\.\d+.\d+)[^\d].*`)

**Comment by schu** [comment]
Hi, thanks for the report. Does the fix in #192 work for you?

**Comment by andreynkravtsov** [comment]
Yes, I validated - that fix also works

**Comment by schu** [comment]
Merged #192 - closing.
